deployment/Environment.md
# Environment

This document provides detailed information about environment variables for each service in Aerie.

- [Aerie Gateway](#aerie-gateway)
- [Aerie Merlin](#aerie-merlin)
- [Aerie Scheduler](#aerie-scheduler)
- [Aerie Sequencing](#aerie-sequencing)
- [Aerie UI](#aerie-ui)
- [Hasura](#hasura)
- [Postgres](#postgres)

## Aerie Gateway

See the [environment variables document](https://github.com/NASA-AMMOS/aerie-gateway/blob/develop/docs/ENVIRONMENT.md) in the Aerie Gateway repository.

## Aerie Merlin

| Name                                  | Description                                                                                                                 | Type      | Default                         |
|---------------------------------------|-----------------------------------------------------------------------------------------------------------------------------|-----------|---------------------------------|
| `AERIE_DB_HOST`                       | The DB instance that Merlin will connect with                                                                               | `string`  | postgres                        |
| `AERIE_DB_PORT`                       | The DB instance port number that Merlin will connect with                                                                   | `number`  | 5432                            |
| `JAVA_OPTS`                           | Configuration for Merlin's logging level and output file                                                                    | `string`  | log level: warn. output: stderr |
| `MERLIN_PORT`                         | Port number for the Merlin server                                                                                           | `number`  | 27183                           |
| `MERLIN_LOCAL_STORE`                  | Local storage for Merlin in the container                                                                                   | `string`  | /usr/src/app/merlin_file_store  |
| `MERLIN_DB_USER`                      | Username of the Merlin DB User                                                                                              | `string`  | merlin_service                  |
| `MERLIN_DB_PASSWORD`                  | Password of the Merlin DB User                                                                                              | `string`  |                                 |
| `UNTRUE_PLAN_START`                   | Temporary solution to provide plan start time to models, should be set to a time that models will not fail to initialize on | `string`  |                                 |
| `ENABLE_CONTINUOUS_VALIDATION_THREAD` | Flag to enable a worker thread that continuously computes and caches activity directive validation results                  | `boolean` | true                            |
| `VALIDATION_THREAD_POLLING_PERIOD`    | Number of milliseconds the above worker thread should wait before querying the database for new, unvalidated directives     | `string`  | 500                             |

## Aerie Merlin Worker

| Name                                     | Description                                                                                                                 | Type     | Default                                      |
|------------------------------------------|-----------------------------------------------------------------------------------------------------------------------------|----------|----------------------------------------------|
| `AERIE_DB_HOST`                          | The DB instance that Merlin will connect with                                                                               | `string` | postgres                                     |
| `AERIE_DB_PORT`                          | The DB instance port number that Merlin will connect with                                                                   | `number` | 5432                                         |
| `JAVA_OPTS`                              | Configuration for Merlin's logging level and output file                                                                    | `string` | log level: warn. output: stderr              |
| `MERLIN_WORKER_LOCAL_STORE`              | The local storage as for the Merlin container                                                                               | `string` | /usr/src/app/merlin_file_store               |
| `MERLIN_DB_USER`                         | Username of the Merlin DB User                                                                                              | `string` | merlin_service                               |
| `MERLIN_DB_PASSWORD`                     | Password of the Merlin DB User                                                                                              | `string` | (this must the same as the Merlin container) |
| `SIMULATION_PROGRESS_POLL_PERIOD_MILLIS` | Cadence at which the worker will report simulation progress to the database.                                                | `number` | 5000                                         |
| `UNTRUE_PLAN_START`                      | Temporary solution to provide plan start time to models, should be set to a time that models will not fail to initialize on | `string` |                                              |

## Aerie Scheduler

| Name                          | Description                                                      | Type     | Default                         |
|-------------------------------|------------------------------------------------------------------|----------|---------------------------------|
| `AERIE_DB_HOST`               | The DB instance that the Scheduler will connect with             | `string` | postgres                        |
| `AERIE_DB_PORT`               | The DB instance port number that the Scheduler will connect with | `number` | 5432                            |
| `HASURA_GRAPHQL_ADMIN_SECRET` | The admin secret for Hasura which gives admin access if used.    | `string` |                                 |
| `JAVA_OPTS`                   | Configuration for the scheduler's logging level and output file  | `string` | log level: warn. output: stderr |
| `MERLIN_GRAPHQL_URL`          | URI of the Merlin graphql interface to call                      | `string` | http://hasura:8080/v1/graphql   |
| `SCHEDULER_DB_USER`           | Username of the Scheduler DB User                                | `string` | scheduler_service               |
| `SCHEDULER_DB_PASSWORD`       | Password of the Scheduler DB User                                | `string` |                                 |
| `SCHEDULER_PORT`              | Port number for the scheduler server                             | `number` | 27185                           |

## Aerie Scheduler Worker

| Name                          | Description                                                           | Type     | Default                                            |
|-------------------------------|-----------------------------------------------------------------------|----------|----------------------------------------------------|
| `AERIE_DB_HOST`               | The DB instance that the Scheduler will connect with                  | `string` | postgres                                           |
| `AERIE_DB_PORT`               | The DB instance port number that the Scheduler will connect with      | `number` | 5432                                               |
| `HASURA_GRAPHQL_ADMIN_SECRET` | The admin secret for Hasura which gives admin access if used.         | `string` |                                                    |
| `JAVA_OPTS`                   | Configuration for the scheduler's logging level and output file       | `string` | log level: warn. output: stderr                    |
| `MERLIN_GRAPHQL_URL`          | URI of the Merlin graphql interface to call                           | `string` | http://hasura:8080/v1/graphql                      |
| `MERLIN_LOCAL_STORE`          | Local storage for Merlin in the container (for backdoor jar access)   | `string` | /usr/src/app/merlin_file_store                     |
| `SCHEDULER_DB_USER`           | Username of the Scheduler DB User                                     | `string` | scheduler_service                                  |
| `SCHEDULER_DB_PASSWORD`       | Password of the Scheduler DB User                                     | `string` |                                                    |
| `SCHEDULER_OUTPUT_MODE`       | How scheduler output is sent back to Aerie                            | `string` | UpdateInputPlanWithNewActivities                   |
| `MAX_NB_CACHED_SIMULATION_ENGINES` | The maximum number of simulation engines to cache in memory during a scheduling run. Must be at least 1 | `number` | 1                                                  |

## Aerie Sequencing

| Name                          | Description                                                   | Type     | Default                            |
|-------------------------------|---------------------------------------------------------------|----------|------------------------------------|
| `AERIE_DB_HOST`               | Hostname of Postgres instance                                 | `string` | postgres                           |
| `AERIE_DB_PORT`               | Port of Postgres instance                                     | `number` | 5432                               |
| `HASURA_GRAPHQL_ADMIN_SECRET` | The admin secret for Hasura which gives admin access if used. | `string` |                                    |
| `LOG_FILE`                    | Either an output filepath to log to, or 'console'             | `string` | console                            |
| `LOG_LEVEL`                   | Logging level for filtering logs                              | `string` | warn                               |
| `MERLIN_GRAPHQL_URL`          | URI of the Aerie GraphQL API                                  | `string` | http://hasura:8080/v1/graphql      |
| `SEQUENCING_DB_USER`          | Username of the Sequencing DB User                            | `string` | sequencing_service                 |
| `SEQUENCING_DB_PASSWORD`      | Password of the Sequencing DB User                            | `string` |                                    |
| `SEQUENCING_LOCAL_STORE`      | Local storage file storage in the container                   | `string` | /usr/src/app/sequencing_file_store |
| `SEQUENCING_SERVER_PORT`      | Port the server listens on                                    | `number` | 27184                              |

## Aerie UI

See the [environment variables document](https://github.com/NASA-AMMOS/aerie-ui/blob/develop/docs/ENVIRONMENT.md) in the Aerie UI repository.

## Hasura

| Name                          | Description                                                   | Type     |
|-------------------------------|---------------------------------------------------------------|----------|
| `AERIE_DATABASE_URL`          | Url of the Aerie Postgres database.                           | `string` |
| `AERIE_MERLIN_URL`            | Url of the Merlin service.                                    | `string` |
| `AERIE_SCHEDULER_URL`         | Url of the scheduler service.                                 | `string` |
| `AERIE_SEQUENCING_URL`        | Url of the sequencing service.                                | `string` |
| `HASURA_GRAPHQL_ADMIN_SECRET` | The admin secret for Hasura which gives admin access if used. | `string` |
| `HASURA_GRAPHQL_JWT_SECRET`   | The JWT secret for JSON web token auth. Also in Gateway.      | `string` |

Additionally, Hasura provides documentation on it's own environment variables you can use to fine-tune your deployment:

1. [graphql-engine](https://hasura.io/docs/latest/graphql/core/deployment/graphql-engine-flags/reference.html#server-flag-reference)
2. [metadata and migrations](https://hasura.io/docs/latest/graphql/core/migrations/advanced/auto-apply-migrations.html#applying-migrations)

## Postgres

The default Aerie deployment uses the default Postgres environment. See the [Docker Postgres documentation](https://hub.docker.com/_/postgres) for more complete information on those environment variables and how to use them.

| Name                     | Description                                            | Type     |
|--------------------------|--------------------------------------------------------|----------|
| `AERIE_USERNAME`         | Username of the Aerie DB User, which owns the Aerie DB | `string` |
| `AERIE_PASSWORD`         | Password of the Aerie DB User, which owns the Aerie DB | `string` |
| `GATEWAY_DB_USER`        | Username of the Gateway DB User                        | `string` |
| `GATEWAY_DB_PASSWORD`    | Password of the Gateway DB User                        | `string` |
| `MERLIN_DB_USER`         | Password of the Merlin DB User                         | `string` |
| `MERLIN_DB_PASSWORD`     | Password of the Merlin DB User                         | `string` |
| `SCHEDULER_DB_USER`      | Username of the Scheduler DB User.                     | `string` |
| `SCHEDULER_DB_PASSWORD`  | Password of the Scheduler DB User                      | `string` |
| `SEQUENCING_DB_USER`     | Username of the Sequencing DB User.                    | `string` |
| `SEQUENCING_DB_PASSWORD` | Password of the Sequencing DB User                     | `string` |

[svelte-kit-adapter-node-docs]: https://github.com/sveltejs/kit/blob/master/packages/adapter-node/README.md


html/_sources/UI-Configurability.md.txt
Users can create custom planning views for different sub-systems    
(e.g. science, engineering, thermal, etc.), where only data (e.g. activities and states) for those sub-systems are visualized. This is done through a mission-authored [JSON](https://www.json.org/json-en.html) file. The format of that file and how to update it is the subject of this document.

# Panel Editor

In order to change the configuration object, navigate to the plan page  
and press `⌘e (command e)`, which should open up the Panel Editor in the right drawer.  
Make your changes to the JSON object and press `⌘s (command s)`.  
You should see the UI update with your changes as long as your JSON object has no errors.  
Check the [console](https://webmasters.stackexchange.com/a/77337) if you suspect your JSON object has errors. If your errors persist, don't hesitate to ask on [#mpsa-aerie-users](https://app.slack.com/client/T024LMMEZ/C0163E42UBF). 

**Note that edited configurations are not currently saved between browser refreshes. This will be added in future releases.**

<img width="1024" src="https://github.jpl.nasa.gov/storage/user/2179/files/6a2d0380-bfa7-11ea-917e-85a7af6047c9">

# Panels

The planning UI consists of a list of panels, and each `Panel` has the following [interface](https://www.typescriptlang.org/docs/handbook/interfaces.html):

```ts
interface Panel {
  bands?: Band[]; // For type 'timeline'
  id: string;
  iframe?: { // For type 'iframe'
    src: string;
  };
  menu?: PanelMenuItem[];
  size: number; // Size percentage of the panel (0% - 100%)
  table?: { // For type 'table'
    columns: string[];
    type: 'activity'; // Only 'activity' typed tables are allowed right now
  };
  title: string;
  type: 'iframe' | 'table' | 'timeline';
}
```

## Menu

Each panel can have an associated menu specified with an `action`, `icon`, and `title`.  
The current actions we support are:

1. `link` adds a link that opens a new browser tab to the specified `url` in the `data` object.
1. `restore` is useful if the panel is type [timeline](https://github.jpl.nasa.gov/MPS/aerie/wiki/UI-Configurability#timeline). It resets the timeline to it's max-time-rage (i.e. zooms all the way out).
1. `simulate` runs a simulation. Any state bands with simulation result states will be updated after a simulation. 

More actions will be supported in the future and they will be more customizable. The icon should be a [material icon](https://material.io/resources/icons/?style=baseline). The menu interface looks like this:

```ts
export type PanelMenuItemAction = 'link' | 'restore' | 'simulate';

export interface PanelMenuItem {
  action: PanelMenuItemAction;
  data?: {
    url?: string;
  };
  icon: string;
  title: string;
}
```

A couple example menu JSON objects looks like this:

```json
[
  {
    "action": "link",
    "data": {
      "url": "https://google.com"
    },
    "icon": "link",
    "title": "Google"
  },
  {
    "action": "restore",
    "icon": "restore",
    "title": "Restore Time"
  }
]
```

## Types

There are currently three types of supported panels: iframe, table, and timeline. The following sections will detail how to create each of these panel types.

### Inline Frame (iframe)

The [iframe](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/iframe) panel allows you to embed a custom HTML page inside of the panel. To create an iframe panel you need to specify `type: "iframe"`, a unique `id`, a `size`, a `title`, and an `iframe` object with a `src`. The `src` is a URL of the HTML page you are embedding in the panel. Here is a basic example of specifying an iframe panel:

```json
{
  "id": "panel3",
  "iframe": {
    "src": "https://www.chartjs.org/samples/latest/charts/line/basic.html"
  },
  "size": 100,
  "title": "Line Chart",
  "type": "iframe"
}
```

### Table

The table panel allows you to view data as a table with columns and rows. You can currently only create table panels for activity data. To create a table panel you need to specify `type: "table"`, a unique `id`, a `size`, a `title`, and a `table` object. The table object specifies the columns you want to see in your data. For example if you want to see an activities `startTimestamp` property you specify it in the column. There is also a special `select` column that allows the column to be selected. Here is a basic example of specifying a table panel:

```json
{
  "id": "panel2",
  "size": 100,
  "table": {
    "columns": [
      "select",
      "type",
      "startTimestamp"
    ],
    "type": "activity"
  },
  "title": "Activity Table",
  "type": "table"
}
```

### Timeline

The timeline panel allows you to specify visualizations of time-ordered data. To create a timeline panel you need to specify `type: "timeline"`, a unique `id`, a `size`, a `title`, and a list of `bands`. Here is the minimum JSON needed to specify a timeline panel:

```json
{
  "bands": [],
  "id": "panel0",
  "size": 100,
  "title": "My First Panel",
  "type": "timeline"
}
```

To visualize data in a timeline you need to add band objects to the `bands` array. A band is a layered visualization of time-ordered data. Each layer of a band is specified as an object of the `subBands` array. The interfaces for a `Band` and `SubBand` are as follows:

```ts
interface Band {
  height?: number;
  id: string;
  subBands: SubBand[];
  yAxes?: Axis[];
}

interface SubBand {
  chartType: 'activity' | 'line' | 'x-range'; // How we actually visualize the data
  color?: string; // Color of all data points
  filter?: {
    activity?: {
      type?: string; // JS regex filtering the type of activity we want to see
    };
    state?: {
      name?: string; // JS regex filtering the name of state we want to see
    };
  };
  id: string;
  type: 'activity' | 'state'; // Type of data the sub-band visualizes
  yAxisId?: string; // Optional y-axis ID link
}
```

Here is a JSON object that creates a single band with one activity sub-band. Notice there are no `yAxes`, as activities do not typically have y-values. Also notice the `filter` property, which is a [JavaScript Regular Expression](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Regular_Expressions) that specifies we only want to see `activity` of `type` `.*`. This is a regex for giving all activity types.

```json
{
  "id": "band0",
  "subBands": [
    {
      "chartType": "activity",
      "filter": {
        "activity": {
          "type": ".*"
        }
      },
      "id": "subBand0",
      "type": "activity"
    }
  ]
}
```

For data that has y-values (for example state data), you can specify a y-axis and link a sub-band to it by ID. Here are the interfaces for `Axis` and `Label`:

```ts
interface Axis {
  id: string;
  color?: string;
  label?: Label;
  scaleDomain?: number[];
  tickCount?: number;
}

interface Label {
  align?: CanvasTextAlign;
  baseline?: CanvasTextBaseline;
  color?: string;
  fontFace?: string;
  fontSize?: number;
  hidden?: boolean;
  text: string;
}
```

Y-axes are specified in the band separately from sub-bands so we can specify multi-way relationships between axes and sub-bands. For example you could have many sub-bands corresponding to a single axis.

Here is the JSON for creating a band with two overlayed `state` sub-bands. The first sub-band shows only states with the name `peel`, and uses the y-axis with ID `yAxis1`. The second sub-band shows only states with the name `fruit`, and uses the y-axis with the ID `yAxis2`.

```json
{
  "id": "band1",
  "subBands": [
    {
      "chartType": "line",
      "filter": {
        "state": {
          "name": "peel"
        }
      },
      "id": "subBand1",
      "type": "state",
      "yAxisId": "yAxis1"
    },
    {
      "chartType": "line",
      "filter": {
        "state": {
          "name": "fruit"
        }
      },
      "id": "subBand2",
      "type": "state",
      "yAxisId": "yAxis2"
    }
  ],
  "yAxes": [
    {
      "id": "yAxis1",
      "label": {
        "text": "peel"
      }
    },
    {
      "id": "yAxis2",
      "label": {
        "text": "fruit"
      }
    }
  ]
}
```

CHANGELOG.md
# Release Notes

#### Features Delivered - December 17th, 2018

| Feature                                                      | Ticket ID                                                    | Description |
| ------------------------------------------------------------ | ------------------------------------------------------------ | ----------- |
| Command Linting, Hinting functionality                       | [MPSSEQ-747](https://cae-jira.jpl.nasa.gov/browse/MPSSEQ-747) |             |
| Capability to Select an Adaptation                           | [MPSSEQ-703](https://cae-jira.jpl.nasa.gov/browse/MPSSEQ-703) |             |
| Component that lists Activity Types                          | [MPSSEQ-701](https://cae-jira.jpl.nasa.gov/browse/MPSSEQ-701) |             |
| Component Displays Activity Constraints                      | [MPSSEQ-700](https://cae-jira.jpl.nasa.gov/browse/MPSSEQ-700) |             |
| Visualize Decomposition of Activity in an Activity Type      | [MPSSEQ-699](https://cae-jira.jpl.nasa.gov/browse/MPSSEQ-699) |             |
| Backend mocks to provide functionality on Activity Types     | [MPSSEQ-615](https://cae-jira.jpl.nasa.gov/browse/MPSSEQ-615) |             |
| Common GUI to navigate through planning, editing and visualization components | [MPSSEQ-608](https://cae-jira.jpl.nasa.gov/browse/MPSSEQ-608) |             |
| Autocomplete parameters in activity editor                   | [MPSSEQ-601](https://cae-jira.jpl.nasa.gov/browse/MPSSEQ-601) |             |
| Componnent to edit values of parameters in an activity instance | [MPSSEQ-587](https://cae-jira.jpl.nasa.gov/browse/MPSSEQ-587) |             |
| List and explore Commands in a Command Dictionary provided   | [AMPSA-19](https://cae-jira.jpl.nasa.gov/browse/AMPSA-19)    |             |
| Edit commands in an editor                                   | [AMPSA-18](https://cae-jira.jpl.nasa.gov/browse/AMPSA-18)    |             |



sequencing-server/src/env.ts
export type Env = {
  HASURA_GRAPHQL_ADMIN_SECRET: string;
  LOG_FILE: string;
  LOG_LEVEL: string;
  MERLIN_GRAPHQL_URL: string;
  PORT: string;
  AERIE_DB_HOST: string;
  AERIE_DB_PORT: string;
  SEQUENCING_DB_USER: string;
  SEQUENCING_DB_PASSWORD: string;
  STORAGE: string;
  SEQUENCING_WORKER_NUM: string;
  SEQUENCING_MAX_WORKER_NUM : string
  SEQUENCING_MAX_WORKER_HEAP_MB: string;
  TRANSPILER_ENABLED: string;
};

export const defaultEnv: Env = {
  HASURA_GRAPHQL_ADMIN_SECRET: '',
  LOG_FILE: 'console',
  LOG_LEVEL: 'info',
  MERLIN_GRAPHQL_URL: 'http://hasura:8080/v1/graphql',
  PORT: '27184',
  AERIE_DB_HOST: 'localhost',
  SEQUENCING_DB_PASSWORD: '',
  AERIE_DB_PORT: '5432',
  SEQUENCING_DB_USER: '',
  STORAGE: 'sequencing_file_store',
  SEQUENCING_WORKER_NUM: '8',
  SEQUENCING_MAX_WORKER_HEAP_MB: '1000',
  SEQUENCING_MAX_WORKER_NUM : '8',
  TRANSPILER_ENABLED: 'true',
};

export function getEnv(): Env {
  const { env } = process;

  const HASURA_GRAPHQL_ADMIN_SECRET = env['HASURA_GRAPHQL_ADMIN_SECRET'] ?? defaultEnv.HASURA_GRAPHQL_ADMIN_SECRET;
  const LOG_FILE = env['LOG_FILE'] ?? defaultEnv.LOG_FILE;
  const LOG_LEVEL = env['LOG_LEVEL'] ?? defaultEnv.LOG_LEVEL;
  const MERLIN_GRAPHQL_URL = env['MERLIN_GRAPHQL_URL'] ?? defaultEnv.MERLIN_GRAPHQL_URL;
  const PORT = env['SEQUENCING_SERVER_PORT'] ?? defaultEnv.PORT;
  const AERIE_DB_HOST = env['AERIE_DB_HOST'] ?? defaultEnv.AERIE_DB_HOST;
  const AERIE_DB_PORT = env['AERIE_DB_PORT'] ?? defaultEnv.AERIE_DB_PORT;
  const SEQUENCING_DB_USER = env['SEQUENCING_DB_USER'] ?? defaultEnv.SEQUENCING_DB_USER;
  const SEQUENCING_DB_PASSWORD = env['SEQUENCING_DB_PASSWORD'] ?? defaultEnv.SEQUENCING_DB_PASSWORD;
  const STORAGE = env['SEQUENCING_LOCAL_STORE'] ?? defaultEnv.STORAGE;
  const SEQUENCING_WORKER_NUM = env['SEQUENCING_WORKER_NUM'] ?? defaultEnv.SEQUENCING_WORKER_NUM;
  const SEQUENCING_MAX_WORKER_NUM = env['SEQUENCING_MAX_WORKER_NUM'] ?? defaultEnv.SEQUENCING_MAX_WORKER_NUM;
  const SEQUENCING_MAX_WORKER_HEAP_MB =
    env['SEQUENCING_MAX_WORKER_HEAP_MB'] ?? defaultEnv.SEQUENCING_MAX_WORKER_HEAP_MB;
  const TRANSPILER_ENABLED = env['TRANSPILER_ENABLED'] ?? defaultEnv.TRANSPILER_ENABLED;
  return {
    HASURA_GRAPHQL_ADMIN_SECRET,
    LOG_FILE,
    LOG_LEVEL,
    MERLIN_GRAPHQL_URL,
    PORT,
    AERIE_DB_HOST,
    AERIE_DB_PORT,
    SEQUENCING_DB_USER,
    SEQUENCING_DB_PASSWORD,
    STORAGE,
    SEQUENCING_WORKER_NUM,
    SEQUENCING_MAX_WORKER_NUM,
    SEQUENCING_MAX_WORKER_HEAP_MB,
    TRANSPILER_ENABLED,
  };
}
